scionlab/tests/test_user_as_models.py

```diff
@@ -123,69 +139,80 @@ def create_and_check_useras(testcase,
-        if overridden_bind_ip:
-            overridden_bind_port = bind_port or public_port
-
-        utils.check_link(testcase, link, utils.LinkDescription(
-            type=Link.PROVIDER,
-            from_as_id=attachment_point.AS.as_id,
-            from_public_ip=_get_public_ip_testtopo(attachment_point.AS.as_id),
-            from_bind_ip=None,
-            from_internal_ip=DEFAULT_HOST_INTERNAL_IP,
-            to_public_ip=public_ip,
-            to_public_port=public_port,
-            to_bind_ip=overridden_bind_ip,
-            to_bind_port=overridden_bind_port,
-            to_internal_ip=DEFAULT_HOST_INTERNAL_IP,
-        ))
-
-    _check_attachment_point(testcase, attachment_point)
+
+    # Check that the AttachmentPoints in `aps_confs` are now AttachmentPoints of the user_as
+    aps_ases = [c.attachment_point.AS for c in aps_confs]
+    user_as_aps_ases = [l.interfaceA.AS
+                        for l in Link.objects.filter(interfaceB__AS=user_as).all()]
+    testcase.assertEqual(set(user_as_aps_ases), set(aps_ases))
+    # Check attachment points configuration
+    for ap_conf in filter(lambda ap_conf: ap_conf.active, aps_confs):
+        ap = ap_conf.attachment_point
+        utils.check_as(testcase, ap.AS)
+        _check_attachment_point(testcase, ap)
+        link = ap_conf.link
+        testcase.assertEqual(ap_conf.public_port, link.interfaceB.public_port)
+
+        if ap_conf.use_vpn:
+            testcase.assertNotEqual(vpn_choice, VPNChoice.NONE)
+            testcase.assertTrue(VPNClient.objects.filter(host=user_as.host,
+                                                         vpn=ap.vpn
+                                                         ).exists())
+            utils.check_link(testcase, link, utils.LinkDescription(
+                type=Link.PROVIDER,
+                from_as_id=ap.AS.as_id,
+                from_public_ip=ap.vpn.server_vpn_ip,
+                from_bind_ip=None,
+                from_internal_ip=DEFAULT_HOST_INTERNAL_IP,
+                to_public_ip=user_as.hosts.get().vpn_clients.get(active=True, vpn=ap.vpn).ip,
+                to_public_port=ap_conf.public_port,
+                to_bind_ip=None,
+                to_bind_port=None,
+                to_internal_ip=DEFAULT_HOST_INTERNAL_IP,
+            ))
+        else:
+            testcase.assertNotEqual(vpn_choice, VPNChoice.ALL)
+            vpn_client = VPNClient.objects.filter(host=user_as.host,
+                                                  vpn=ap_conf.attachment_point.vpn
+                                                  ).first()
+            testcase.assertTrue(not vpn_client or not vpn_client.active)
+            bind_ip, bind_port = ap_conf.bind_ip, ap_conf.bind_port
+            if installation_type == UserAS.VM:
+                bind_ip = '10.0.2.15'
+                if bind_port is None:
+                    # Port is assigned automatically in this case, so we cannot know it in advance
```

Comment:
Why? Why is this not just the public port, as it seems to have previously been? I think this particular snippet here, determining the "overriden bind ip/port", should not require any changes.

Reply:
If I'm not mistaken, the old 'overridden bind ip/port' does not always work.
In fact, looking at the `InterfaceManager:create(...)` method, if the `bind_ip` is specified, and the `bind_port` is not, the latter is chosen based on the `Portmap:get_port(...)` method. Now, it's true that the `preferred` port to be chosen is the `public_port`, but it might happen that the `public_port` is already taken, and thus the `bind_port` will be different.